databuilder/query_model/nodes.py

```diff
@@ -378,9 +378,15 @@ class Case(Series[T]):
     default: Optional[Series[T]] = None
 
     def __hash__(self):
-        # `cases` is a dict and so not hashable by default, but we treat it as
-        # immutable once created so we're fine to make it hashable
-        return hash((tuple(self.cases.items()), self.default))
+        # `cases` is a dict and so not naturally hashable, but we treat it as immutable.
+        # The specfic hash used below is based on a recommendation by Raymond Hettinger.
+        # It's almost certainly unnecessary to include the values and default in the
+        # hash, but given that it's all cached anyway we may as well save ourselves the
+        # trouble of thinking about it. See:
+        # https://stackoverflow.com/a/16162138
+        return hash(
+            (frozenset(self.cases), frozenset(self.cases.values()), self.default)
```

Comment:
It might save a rontosecond of reading time if this was `(frozenset(self.cases.keys()), frozenset(self.cases.values()), self.default)`.

Reply:
It wouldn't be Hettinger-approved though! I haven't looked enough at the C implementation of `dict` to know if it would make any difference but he does mention that fact the his approach allows `frozenset` to re-use the key hashes without having to recalculate them.